### DISCUSSION TITLE
Refresh imported folder structure

### DISCUSSION THREAD

**Discussion by DanielFlaum** [freeplane/freeplane#922] [Ideas]
It would be helpful to be able to refresh imported folder structures to reflect files that have been created or removed since the last import. Bonus points for a checkbox that allows changes to be displayed by marking newly created files with one styles and deleted files with another. Yes, there are tools like Meld or Kaleidoscope that can diff folder structures, but they don't lay things out as a mind map that can be edited, annotated, and printed.

I think this could made possible by modifying the existing importer to add appropriate attributes to nodes. The node on which the folder structure was appended could have an attribute indicating the location in the file system that was chosen when the importer was first run. Then, subnodes appended by the importer could have an attribute indicating the file or folder they represent, so that later runs of the importer can detect whether they have been deleted, or whether newly created files need to be updated.

I'm not familiar with Freeplane's source code at all, but I think the place where changes would have to be made is https://github.com/freeplane/freeplane/blob/b76798aef3a47fd1a3f766391e111035e0041199/freeplane/src/main/java/org/freeplane/features/url/mindmapmode/ImportFolderStructureAction.java

**Comment by grotesque** [comment]
https://stackoverflow.com/questions/5942480/mind-mapping-based-on-file-system

Using a mind map to explore/navigate, edit (delete, rename, add files and folders) and annotate a folder structure (adding notes to nodes) is a great feature in my opinion. There are others with the same opinion (see stackoverflow link).

I hope in future Freeplane can replace file explorer for navigation. It can also replace file explorer for file operations. If at the same time it be given the ability to annotate the hierarchy then it can do all this while not ignoring its original purpose: knowledge management.

Below extension may do the above. But it's not working.
https://docs.freeplane.org/attic/old-mediawiki-content/Information_management_with_Freeplane_add-on_dView(indexfile).html

This might work:

https://sourceforge.net/p/freeplane/discussion/758437/thread/f25fa9a068/

**Reply by EdoFro** [reply]
> This might work:
> 
> https://sourceforge.net/p/freeplane/discussion/758437/thread/f25fa9a068/

Yes, it works. I use it daily. Almost all my maps use this feature.

Sadly I created these scripts before learning how to pack an add on, so it is a little cumberstone to install.
 Someday I'm going to transform it to an add-on . It is in my pending tasks list.

From all the scripts and add-ons that I have created, these are the most helpful to me and my day to day work.

greetings,
edo

**Reply by grotesque** [reply]
I wasn't able to make it work on new mind map. It works on your sample mind map. And by work I mean it does something but I cannot figure out how that something helps. Maybe I need to find good english documentation to understand the feature.

**Reply by bepolymathe** [reply]
> Someday I'm going to transform it to an add-on . It is in my pending tasks list.

That’s would be great !

**Comment by Illegit** [comment]
I confirm even I use EdoFro MDI daily, fundamental!
note that it doesn't load the file if you have the same file name (and so on)
I use to move files from a folder to a specific folder in a specific folder structure and even changing the folder tree in that map reflects a new folder structure in your pc.

**Comment by EdoFro** [comment]
Just for your information, I started the process of improving MDI.
I released a first version as an add-on for Freeplane and enabled the discussion forum in the project.
I will be posting there all improvements.

project: https://github.com/EdoFro/Freeplane_Map-Drive-Inator

documentation will follow

**Comment by Illegit** [comment]
Tested and working. I'll complete some tests this weekend.
Thanks

**Comment by biswajit-goswami** [comment]
Hi All, 
How do we install this add on?
I tried adding 2 of the .mm files from the git repo.....but FP 1.11.x version always failed. so any help to install teh add on will be great.

REgards,
Biswajit

**Reply by EdoFro** [reply]
you have to download the add-on from its release page in the git repo:

https://github.com/EdoFro/Freeplane_Map-Drive-Inator/releases/latest

**Reply by EdoFro** [reply]
you also need to install another little add-on that installs a library that some of my add-ons use.

look [this message](https://github.com/EdoFro/Freeplane_Map-Drive-Inator/discussions/10#discussioncomment-4723003)

**Reply by biswajit-goswami** [reply]
Thank you @EdoFro  , I was able to install both.
I tried to follow some of the steps..and was able to import the folder structure. Then I made a change - created a new sub folder and moved a file to it..in Windows explorer..I now would like to refresh the folder structure in the mind map accordingly..so I click on the following:
![image](https://github.com/freeplane/freeplane/assets/11291623/374c3090-2518-4356-858e-2a3cca3e12d1)

But did not see any change in the mind map..am I missing something?

**Reply by EdoFro** [reply]
click on ‘MapDriveInator’ command

**Reply by EdoFro** [reply]
that command ‘coordinates/updates/reflects’ changes from drive to map and from map to drive.

**Reply by biswajit-goswami** [reply]
Thanks @EdoFro  , it works.

**Comment by biswajit-goswami** [comment]
@EdoFro , if I need to skip some sub folders during the import , is it possible? Currently, if there is a file matching the filter, the parent folder is getting  included.....if I delete it from the map, will it get deleted from the file system too?

**Reply by EdoFro** [reply]
Hi @biswajit-goswami 

> if I delete it from the map, will it get deleted from the file system too?

No, MDI doesn't delete any file from disk. I decided not to put that feature in the add-on.

What I do:
I create a "delete" folder, and, when I want to delete a file in the mindmap, I move that node to that folder. Later, in Windows Explorer, I delete all the files in that folder.


> if I need to skip some sub folders during the import , is it possible?

Yes, you have to add the "locked" style to the node. Any folders or files under a "locked" node will be skipped.

![image](https://github.com/freeplane/freeplane/assets/35700575/c15045d4-3f7b-4046-b2f5-c050df4e3e4c)


Hope this helps,
edo

**Reply by biswajit-goswami** [reply]
@EdoFro  , I tried to import the folders (not files) for a folder as shown below , after creating the base folder node:
![image](https://github.com/freeplane/freeplane/assets/11291623/26cf8d2b-afa7-4ead-8869-4b00d5d7379c)
![image](https://github.com/freeplane/freeplane/assets/11291623/20954218-6be8-4264-b243-522bd00ec4b8)

But the structure I got was not completely right:
![image](https://github.com/freeplane/freeplane/assets/11291623/1ea89c29-83fe-46a7-a7aa-ab72086f028f)

 As can be seen above , the folder node b and c are mapped under two copies of node a ..is this expected?
 
 However, if I skip this step, and import all the files , then the folder structure is correct.
Any inputs?

**Reply by EdoFro** [reply]
I transformed this in a new issue in its repository:
https://github.com/EdoFro/Freeplane_Map-Drive-Inator/issues/40

**Reply by biswajit-goswami** [reply]
I have added the details in the issue.

**Comment by EdoFro** [comment]
I created a sort of quick starting guide for MDI here: https://github.com/EdoFro/Freeplane_Map-Drive-Inator/discussions/39
